int32 good_angle
int32 far_good_angle
int32 final_angle
int32 dd_1
int32 dd_2
int32 df_1
int32 df_2
